### PR TITLE
build(deps) update Rust dependencies (2023-W51)

### PR DESCRIPTION
**Pull request by fnichol**
This was a large update, primarily due to a longer than usual gap between crate updates. During this period we were adopting/porting our build system over to Buck2 and then deploying our CI/CD system so this task fell a little by the wayside. No harm, but this should not be our default mode of operation 😉 

What's new or of note?

- **Rust updated to 1.74.1**: it appears as though our build time regression of `lib/sdf-server` has been resolved and this brings us back to the stable release train.
- **Upgrade to ring 0.17.5**: while not normally a big callout, this crate was very complex to build in Buck2. Thankfully the Meta maintainers updated their Reindeer fixups to `0.17.5` from the `0.16.x` stream. This had a lot of breaking changes in the build side and as future version bumps of this crate may require us to revisit the fixups, we opted to pin this crate for now, just like in the Buck2 project upstream.
- **Upgrades to all telemetry crates**: all the `tracing-*` and `opentelemetry-*` crates are up to date and work with each other. These dependencies were all pinned back while waiting for some pending releases which have long since shipped.
- **Upgrade to current async-nats crate**: the biggest change here is the client has a first-class struct of `Subject` for NATS subjects (rather than a generic `String` type in past releases). This felt natural and useful, so this change was threaded out to our calling code.

What's outstanding?

In short: http/hyper 1.x releases. There is a lot of change here and it may impact upgrading `axum`, `tower-http`, and `tokio-tungstenite`. As a result, we'll do this in the new year.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -45,10 +45,10 @@ members = [
 ]
 
 [workspace.dependencies]
-async-nats = "0.31.0"
+async-nats = "0.33.0"
 async-recursion = "1.0.4"
 async-trait = "0.1.68"
-axum = { version = "0.6.18", features = ["macros", "multipart", "ws"] }
```
Comment (nickgerace) [inline Cargo.toml:51]: Good call

File: Cargo.toml [hunk 2]
```diff
@@ -59,42 +59,43 @@ clap = { version = "4.2.7", features = ["derive", "color", "env", "wrap_help"] }
...
 crossbeam-channel = "0.5.8"
-deadpool = { version = "0.9.5", features = ["rt_tokio_1"] }
-deadpool-postgres = "0.10.5"
+deadpool = { version = "0.10.0", features = ["rt_tokio_1"] }
+deadpool-postgres = "0.12.1"
 derive_builder = "0.12.0"
 derive_more = "0.99.17"
 diff = "0.1.13"
 directories = "5.0.1"
-docker-api = { "git" = "https://github.com/vv9k/docker-api-rs.git", rev = "b1f1891d1d17cb522cc20212452c8c039c5e0810" }
```
Comment (nickgerace) [inline Cargo.toml:72]: Nice!

File: Cargo.toml [hunk 3]
```diff
@@ -132,27 +133,32 @@ tokio-postgres = { version = "0.7.8", features = ["runtime", "with-chrono-0_4",
...
 [patch.crates-io]
 # pending a potential merge and release of
 # https://github.com/softprops/hyperlocal/pull/53
 hyperlocal = { git = "https://github.com/fnichol/hyperlocal.git", branch = "pub-unix-stream" }
 # pending a potential merge and release of
-# https://github.com/mitsuhiko/self-replace/pull/18
-self-replace = { git = "https://github.com/systeminit/self-replace.git", branch = "unix-current-exe-read-link" }
+# https://github.com/vv9k/docker-api-rs/issues/69
+# See: https://github.com/systeminit/si/pull/2903
+docker-api = { git = "https://github.com/vv9k/docker-api-rs.git", branch = "master" }
```
Comment (nickgerace) [inline Cargo.toml:160]: Nice part 2

File: lib/council-server/src/client.rs [hunk 4]
```diff
@@ -1,5 +1,5 @@
 use futures::StreamExt;
-use si_data_nats::{NatsClient, Subscriber};
```
Comment (nickgerace) [inline lib/council-server/src/client.rs:2]: Gotcha. `async-nats` change.

File: lib/sdf-server/tests/service_tests/crdt.rs [hunk 5]
```diff
@@ -41,8 +41,8 @@ async fn client(doc: Doc, server: &Server) -> Result<Client, Box<dyn std::error:
 
     let (shutdown_broadcast_tx, shutdown_broadcast_rx) = broadcast::channel(1);
 
-    let subscription = server.nats.subscribe(&server.channel_name).await?;
-    let ws_subscription = server.nats.subscribe(&server.channel_name).await?;
+    let subscription = server.nats.subscribe(server.channel_name.clone()).await?;
```
Comment (nickgerace) [inline lib/sdf-server/tests/service_tests/crdt.rs:44]: Oh this is going to hurt everywhere, isn't it?

File: support/buck2/buck2-sync-cargo-deps.py [hunk 6]
```diff
@@ -81,6 +91,86 @@ def main() -> int:
     return 0
 
 
+def update_cargo_lock(
```
Comment (nickgerace) [inline support/buck2/buck2-sync-cargo-deps.py:94]: Great fix

File: support/buck2/buck2-sync-cargo-deps.py [hunk 7]
```diff
@@ -81,6 +91,86 @@ def main() -> int:
...
+            "diff",
+            "-u",
+            backup_lock.name,
+            cargo_lock_file,
+        ]
+        result = subprocess.run(cmd, capture_output=True)
+        if result.returncode == 1:
+            print("  - Regenerated {} has changed with diff:\n".format(
+                cargo_lock_file))
+            print(result.stdout.decode("ascii"))
```
Comment (nickgerace) [inline support/buck2/buck2-sync-cargo-deps.py:149]: You have to specify ascii?? ouch.

**Review by nickgerace — APPROVED** [review]

**Comment by fnichol** [comment]
bors merge
